presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketFunction.java

```diff
@@ -14,33 +14,71 @@
 package com.facebook.presto.hive;
 
 import com.facebook.presto.common.Page;
+import com.facebook.presto.common.type.Type;
 import com.facebook.presto.spi.BucketFunction;
 import org.apache.hadoop.hive.serde2.typeinfo.TypeInfo;
 
 import java.util.List;
-import java.util.stream.Collectors;
+import java.util.Optional;
 
+import static com.facebook.presto.hive.HiveBucketProperty.BucketingType.HIVETYPE;
+import static com.facebook.presto.hive.HiveBucketProperty.BucketingType.PRESTOTYPE;
+import static com.facebook.presto.hive.HiveBucketing.getHiveBucket;
 import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.String.format;
 import static java.util.Objects.requireNonNull;
 
 public class HiveBucketFunction
         implements BucketFunction
 {
     private final int bucketCount;
-    private final List<TypeInfo> typeInfos;
+    private final Optional<List<TypeInfo>> typeInfos;
+    private final Optional<List<Type>> types;
+    private final HiveBucketProperty.BucketingType bucketingType;
 
-    public HiveBucketFunction(int bucketCount, List<HiveType> hiveTypes)
+    public static HiveBucketFunction createHiveCompatibleBucketFunction(
+            int bucketCount,
+            List<HiveType> hiveTypes)
+    {
+        return new HiveBucketFunction(bucketCount, Optional.of(hiveTypes), Optional.empty());
+    }
+
+    public static HiveBucketFunction createPrestoNativeBucketFunction(
+            int bucketCount,
+            List<Type> types)
+    {
+        return new HiveBucketFunction(bucketCount, Optional.empty(), Optional.of(types));
+    }
+
+    private HiveBucketFunction(
```

Comment:
Same comments as for other classes. Let's pass `bucketingType` as a parameter, and let's validate if the `hiveTypes` and `types` are set accordingly.